File: Platforms/Common/Reign_Video_D3D11_Component/States/BlendState.cpp
#include "pch.h"
#include "BlendState.h"

namespace Reign_Video_D3D11_Component
{
	#pragma region Constructors
	BlendStateError BlendStateCom::Init(VideoCom^ video, BlendStateDescCom^ desc)
	{
		null();
		this->video = video;

		ID3D11BlendState* stateTEMP;
		if (FAILED(video->device->CreateBlendState(desc->desc, &stateTEMP)))
		{
			return BlendStateError::BlendState;
		}
		state = stateTEMP;

		return BlendStateError::None;
	}

	BlendStateCom::~BlendStateCom()
	{
		if (state) state->Release();
		null();
	}

	void BlendStateCom::null()
	{
		state = 0;
	}
	#pragma endregion

	#pragma region Methods
	void BlendStateCom::Enable()
	{
		ID3D11BlendState* stateTEMP = state;
		video->deviceContext->OMSetBlendState(stateTEMP, 0, 0xffffffff);
	}
	#pragma endregion
}

File: Platforms/Common/Reign.Video.D3D11/States/BlendState.cs
using System;
using Reign.Core;
using Reign_Video_D3D11_Component;

namespace Reign.Video.D3D11
{
	public class BlendState : Disposable, BlendStateI
	{
		#region Properties
		private BlendStateCom com;
		#endregion

		#region Constructors
		public BlendState(DisposableI parent, BlendStateDescI desc)
		: base(parent)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();

				com = new BlendStateCom();
				var error = com.Init(video.com, ((BlendStateDesc)desc).com);

				if (error == BlendStateError.BlendState) Debug.ThrowError("BlendState", "Failed to create BlendState");
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public void Enable()
		{
			com.Enable();
		}
		#endregion
	}
}